### PR TITLE
Implement the ManageSieve protocol (RFC 5804)

### PR DESCRIPTION
**Pull request by monnerat**
This should be rather complete, as it introduces support in library, cli tool, tests and documentation.
Currently marked as experimental.

### CODE REVIEW

**Comment by jay** [comment]
I've never heard of it. Are there many users?

**Comment by bagder** [comment]
I wish you'd discussed this with us before you implemented all this. I've not seen a single user ask for this protocol, ever - and we ask users every year and give them the chance to specify which protocols they miss. I'm not convinced it is a suitable protocol for curl.

Please help us understand why curl users want this esoteric protocol supported?

**Comment by monnerat** [comment]
> I've never heard of it. Are there many users?

I don't know: however there are many users asking for the feature in mail clients.
This is THE protocol allowing a per-user control of mail redirection, auto-replies, spam filtering, server-side filing, etc. There are no other standard for this purpose.
As examples, the PigeonHole module of Dovecot implements such a server, the Horde web groupware uses this protocol and there are Thunderbird client addons for it (https://addons.thunderbird.net/en-US/thunderbird/addon/sieve/).
Existing client implementations: http://sieve.info/libraries#managesieve_the_transport_protocol_for_script_management

>  I've not seen a single user ask for this protocol, ever

Not true: it has been reported once or twice in the surveys, and this where I've seen it mentioned for the first time. I use it regularly on my own mail server since then.

> Please help us understand why curl users want this esoteric protocol supported?

First, it is not more esoteric than dict or gopher (nowadays) and much more useful.

- It completes the mail-oriented protocols we already support.
- Sieve scripts are files, although not general files like with ftp, http, etc. but we already support such files (ldap, dict, rtsp).
- Featuring it in curl will promote its use, and thus the target features mentioned above. Probably many potential client developers have been discouraged implementing it because of the narrow count of support library.
- Its availability in the cli tool also allows checking the script syntax by the server via a single line command (`curl -X checkscript -T script ...`).
- Using sieve scripts allows to do without some well known proprietary server more easily.

> I'm not convinced it is a suitable protocol for curl.

I thought you would be happy with this code. Anyway, you can still refuse it.
I don't expect this protocol have a high score in the future surveys, but its availability will be much valuable.
Did you ever use sieve scripts? You should give them a try!
I hope I've been able to convinced you.

**Comment by bagder** [comment]
> it has been reported once or twice in the surveys, and this is where I've seen it mentioned for the first time

I stand corrected. I went back and indeed I found a request for it in 2018. (didn't check further back)

> First, it is not more esoteric than dict or gopher (nowadays) and much more useful.

1. Don't compare to existing implementations as we might not have accepted those protocols if someone brought them today: https://curl.se/dev/new-protocol.html#dont-compare
2. "more useful" is just an opinion, it's not a fact.

> Did you ever use sieve scripts?

Never. I didn't even know what the protocol did before your PR here...

**Comment by monnerat** [comment]
> I didn't even know what the protocol did before your PR here

One more thing: sieve is the missing link between an SMTP/IMAP-based server and the proprietary server mentioned above...

**Comment by monnerat** [comment]
Not directly related, but discovered while testing/debugging sieve:

Forgive me if I'm wrong, but I strongly suspect W$ CI jobs on azure pipeline to use a wrong version of the library for tests:

- sieve is enabled: confirmed by the configure log.
- sieve.c is compiled: seen in make log
- cli tool is correct: else test 1462 would have failed.
- test 3100 for sieve fails with `RUN: SIEVE server failed verification` and `* Protocol "sieve" not supported or disabled in libcurl` in sieve_verify.log.

I attempted to see what's wrong by updating the `failf()` message in `findprotocol()`, but the original message still appeared as above. As a consequence, I can infer a wrong library is loaded at run-time.

See logs of CI job `windows msys2_mingw64_debug_openssl` in this PR, for example.

As I'm a specialist neither in CI jobs nor in W$ scripting, I would prefer someone else investigate it. If you want me to create a separate issue for that (although not a curl but an infrastructure problem), just ask.

File: .azure-pipelines.yml [hunk 1]
```diff
@@ -203,4 +203,4 @@ stages:
       displayName: 'test'
       env:
         AZURE_ACCESS_TOKEN: "$(System.AccessToken)"
-        TFLAGS: "-vc /usr/bin/curl.exe -r -rm $(tests)"
+        TFLAGS: "-r -rm $(tests)"
```
Comment (mback2k) [inline .azure-pipelines.yml:206]: This change seems to be unrelated.

Comment (monnerat) [reply .azure-pipelines.yml:206]: This is why it is in a temporary commit: this commit allows me to enable functional CI tests for the PR. It is intended to be removed before pulling to master.

Comment (mback2k) [reply .azure-pipelines.yml:206]: Actually I think you are right about removing that parameter and if you would remove it while merging to master, the ManageSieve tests would be disabled again, because the verification of the test server would not work anymore.

Comment (monnerat) [reply .azure-pipelines.yml:206]: Yes, that's it! At least until /usr/bin/curl.exe supports ManageSieve.
That said, I don't think it impacts the test results, as an unsupported protocol seems not be counted as a failure. ManageSieve will not be tested, that's all.
In the current PR, I really need it to be tested: that's why I made the temp change.
I did not follow why this `--vc` option has been enabled, but I can imagine it will force performing tests even if the tested version is not able to verify server.
This option should probably be kept, as we do not implement a new protocol every day :-)

Comment (mback2k) [reply .azure-pipelines.yml:206]: I added that parameter some time ago, because the curl given there will also be used to communicate with the Azure Pipelines test results API and a broken build would cause the communication with the API to fail.

Comment (monnerat) [reply .azure-pipelines.yml:206]: I see. So let's keep it.
Paradoxically the real problem here is the tests came too soon after a feature as been implemented !

Comment (bagder) [reply .azure-pipelines.yml:206]: `-vc /usr/bin/curl.exe` should only be done while debugging, we shouldn't have it the CI all the time because of exactly the reason @monnerat found out: the old version might not do the thing the current test script wants.

Comment (monnerat) [reply .azure-pipelines.yml:206]: @mback2k : as you mastered the Azure Pipeline and I'm not very skilled about it, I prefer to let you resolve this (outside this unrelated PR,  of course!).

**Comment by vszakats** [comment]
Haven't heard of ManageSieve yet, but Sieve is the standard language for creating email-server-side filtering rules. It's quite useful and even though the language is awkward, it's better than proprietary declarative files or nothing at all.

1. Is someone aware of a compilation of email providers / server software supporting ManageSieve?
2. Unless I'm missing something the only way for a secure ManageSieve connection is via `STARTTLS` _upgrade_. This is vulnerable to (trivial by a MitM) downgrade attacks. From security/privacy angles, this makes the protocol problematic. Was this perhaps addressed in an RFC update?

**Comment by monnerat** [comment]
> ...the language is awkward...

It is derived from ACAP: https://www.rfc-editor.org/info/rfc2244.

> Is someone aware of a compilation of email providers / server software supporting ManageSieve?

The protocol has been specified by the Cyrus-IMAP designers.
I did not find a ManageSieve-avare server list, but https://en.wikipedia.org/wiki/Comparison_of_mail_servers#Feature_comparison mentions those supporting sieve scripts. Hopefully the protocol should go with.

> Unless I'm missing something the only way for a secure ManageSieve connection is via STARTTLS upgrade. This is vulnerable to (trivial by a MitM) downgrade attacks. From security/privacy angles, this makes the protocol problematic. Was this perhaps addressed in an RFC update?

There is only https://www.rfc-editor.org/info/rfc7817 addressing TLS, but it only contains server identity checking rules for several mail protocols and does not affect the protocols themselves. I can't find a definition for an implicit SSL enabled service (no sieves scheme).

**Comment by vszakats** [comment]
@monnerat: Thanks for the pointers!

The Sieve table column is definitely useful, though I had _ManageSieve support_ in mind specifically. AFAIU the former doesn't imply the latter. At least the few services I happen to know with Sieve support, do not offer ManageSieve.

As for TLS verification, it's a must indeed, but what if the connection just never gets upgraded to use TLS in the first place? This can happen either because the server response gets manipulated to indicate it isn't supported, or because the client request is manupiulated to never make the STARTTLS request. If this happens without the client noticing and a way to abort, the rest of the session will continue in cleartext, including the authentication phase, content of read filters and even the option for the MitM to inject commands to modify them. With the latter e.g. all future incoming email traffic can be forwarded to a mailbox controlled by the MitM.

**Comment by monnerat** [comment]
> AFAIU the former doesn't imply the latter.

Yes, but we can logically hope it's not the case.

> At least the few services I happen to know with Sieve support, do not offer ManageSieve.

This seems weird: how they allow controlling per-user scripts? Maybe they do feature it, but it's not enabled...

> As for TLS verification, it's a must indeed,...

RFC 7817 also targets SMTP, IMAP and POP3. As we support these protocols, If these rules are not applied by our SSL subsystem, then these protocols are also affected. But this is an SSL problem, not a protocol one.
In curl, we have options `--ssl` and `--ssl-reqd`: the latter should prevent the cases you mention. Use of the former or a clear connection is left at the user's discretion.
Note also that Dovecot does not fallback to clear mode if the TLS negotiation fails: it just disconnects; you cannot then expect to fallback to clear mode... but that's jjust a particular case.
Dovecot also do not offer cleartext authentication methods until TLS has been negotiated. New capabilities are sent to the cliient after successful `STARTTLS`, including an updated set of supported SASL mechanisms.
Commands dealing with scripts/lists are not recognized until authenticated. 

I think the only efficient protection could be made at the authentication level with a new option enabling/disabling cleartext mechanisms when no TLS has been negotiated. But again, that applies to all protocols that can upgrade to TLS, and mainly to the SASL support, but that's another story! For now, `--ssl-reqd` and/or `--login-options` should be used.

**Comment by vszakats** [comment]
`STARTTLS` is just as an issue for already supported email protocols, I agree, but, one difference is that those all support implicit TLS. With ManageSieve there is no secure alternative, even though the managed data is just as sensitive/private. This may be a sign that the protocol isn't actively maintained.

[ The services I know allow editing Sieve filters via their filter configuration web UI. ]

**Comment by monnerat** [comment]
>  The services I know allow editing Sieve filters via their filter configuration web UI.

This is what Horde groupware module ingo does, but it uses ManagesSieve at the server level to interact with the MDA.

File: .azure-pipelines.yml (hunk 1 above)
Comment (mback2k) [reply .azure-pipelines.yml:206]: Will do, see https://github.com/curl/curl/pull/7591.

**Comment by monnerat** [comment]
> one difference is that those all support implicit TLS. With ManageSieve there is no secure alternative

I don't think `--ssl-reqd` is less secure than implicit TLS.
The only potential problem I found is a `REFERRAL` response code in the server greeting message, that could be forged by a MitM to redirect requests to its own server. To avoid this situation, I have now slightly changed the code to refuse following a referral if TLS is requested and has not yet been successfully negotiated, even with `--location` option.

**Comment by vszakats** [comment]
New paper on STARTTLS vulnerabilities: https://www.feistyduck.com/bulletproof-tls-newsletter/issue_80_vulnerabilities_show_fragility_of_starttls

**Comment by monnerat** [comment]
> New paper on STARTTLS vulnerabilities: https://nostarttls.secvuln.info/ [paper] [via]

Thanks for the link. Unfortunately the test tool used in this study does not include ManageSieve (targets SMTP/IMAP/POP3 only) and is written in 3 different languages :-(
There's no `PREAUTH` in ManageSieve, thus we won't have this problem. Among the capabilities transferred in the greetings response, only `STARTTLS` and `SASL` are handled. In addition, a successful TLS negotiation is followed by a new set of capabilities, effectively redefining supported SASL mechanisms. The `REFERRAL` problem has already been handled.
Client command pipelining is not under our control: this is a server business. However we can check for server response pipelining: this is now done.

As the paper says, STARTTLS is fragile in terms of implementation, but I don't see any vulnerability in the concept.

Anyway in our position, we can only be vigilant in our STARTTLS use, but we can't change a protocol definition.

File: lib/sieve.c [hunk 2]
```diff
@@ -0,0 +1,2066 @@
...
+    return result;
+  }
+
+  switch(n) {
+  case 1:       /* sieve://authority/[scriptname] */
+    sieve->scriptname = components[0];
+    break;
+  case 2:       /* sieve://authority/[owner]/[scriptname] */
+    sieve->owner = components[0];
+    sieve->scriptname = components[1];
```
[the remark this answers is not in the document]
Comment (monnerat) [reply lib/sieve.c:320]: @sonatype-lift ignore

**Comment by monnerat** [comment]
I did a quick search for server-side ManageSieve support:

Servers supporting ManageSieve:
- Apache James          https://james.apache.org/
- Cyrus IMAP            https://www.cyrusimap.org/3.4/imap/reference/admin/sieve.html
- Dovecot                       https://wiki1.dovecot.org/ManageSieve
- Kolab                 https://docs.kolab.org/installation-guide/preparing-the-system.html#system-firewall

Servers supporting sieve, but not mentioning ManageSieve, hence it's very unlikely they support it: 
- Axigen
- Courier mail server   
- Kerio Connect
- Mailsite
- Zimbra

Citadel used to support ManageSieve until july 2020, but they then even removed sieve support.

File: lib/sieve.c [hunk 3]
```diff
@@ -0,0 +1,2074 @@
...
+
+  items += itemcount - 1;
+
+  if(items->type == SIEVE_ITEM_STRING) {
+    const char *ptr = (const char *) Curl_bufref_ptr(&items->str);
+    size_t len = Curl_bufref_len(&items->str);
+
+    if(len) {
+      if(warnings || len >= CURL_ERROR_SIZE || !can_quote(ptr, len)) {
+        result = client_write(data, NULL, NULL, ptr, len, NULL);
```
[the remark this answers is not in the document]
Comment (monnerat) [reply lib/sieve.c:1569]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (monnerat) [reply lib/sieve.c:1569]: @sonatype-lift ignore

File: docs/libcurl/opts/CURLOPT_SERVICE_NAME.3 [hunk 4]
```diff
@@ -32,15 +32,15 @@ CURLcode curl_easy_setopt(CURL *handle, CURLOPT_SERVICE_NAME, char *name);
...
+service names are "ftp", "HTTP", "imap", "ldap", "pop", "sieve" and "smtp".
+This option allows you to change them.
 
 The application does not have to keep the string around after setting this
 option.
 .SH DEFAULT
 See above
 .SH PROTOCOLS
-HTTP, FTP, IMAP, LDAP, POP3 and SMTP
+HTTP, FTP, IMAP, LDAP, ManageSieve, POP and SMTP
```
Comment (danielgustafsson) [inline docs/libcurl/opts/CURLOPT_SERVICE_NAME.3:43]: Shouldn't this still say POP3?

Comment (monnerat) [reply docs/libcurl/opts/CURLOPT_SERVICE_NAME.3:43]: Yes it should. Probably an editing glitch.
Thanks for pointing this out.

**Comment by bagder** [comment]
@monnerat, thanks for your long and hard work on this but I don't see support for this PR in the community or among maintainers. I think it is time to drop this and let it go. Focus your efforts on something else.

**Comment by danielgustafsson** [comment]
> Well, you can then close the PR, but I'm afraid I'll still maintain my branch as I'm using it.

I've long had a wishy-washy idea that it would be cool to support protocol plugins/extensions for protocols that isn't suitable for core (ie runtime pluggable, not recompile-with-my-C-file pluggable). This seems like a good driver for such an idea. @bagder has this been discussed over the years?

**Comment by bagder** [comment]
> it would be cool to support protocol plugins/extensions

It has been mentioned and discussed briefly several times over the years, but it has never materialized. I don't think anyone has pushed for it for real and actually tried to make it. Put very simple, it would probably mean a way for an application to provide a protocol handler of its own to a running libcurl.

**Comment by bagder** [comment]
In [the 2024 survey](https://daniel.haxx.se/blog/2024/06/17/curl-user-survey-2024-analysis/), only 3.4% said they want support for the ManageSieve protocol in curl.
